File: src/rieger859663/Image.java
package rieger859663;

import cgtools.Vec3;
import cgtools.ImageWriter;
import java.io.IOException;

public class Image {
    int width;
    int height;
    double imageData[];
    
    public Image(int width, int height) {
        this.width = width;
        this.height = height;
        imageData = new double[width * height * 3];
    }

    public void setPixel(int x, int y, Vec3 color) {
        imageData[y * 3 * width + x * 3    ] = color.x;
        imageData[y * 3 * width + x * 3 + 1] = color.y;
        imageData[y * 3 * width + x * 3 + 2] = color.z;
    }

    public void write(String filename) throws IOException {
        ImageWriter writer = new ImageWriter( imageData, width, height );
        writer.write( filename );
    }
}

File: src/cgtools/image/image.go
package image

import (
    "customtools/vec3"
    "cgtools/imageWriter"
)

const gamma = 2.2

type Image struct {
    width int
    height int
    imageData []float64
}

func New(w int, h int) Image {
    return Image{
        width: w,
        height: h,
        imageData: make([]float64,w * h * 3),
    }
}

func (img *Image) SetPixel(x int, y int, color vec3.Vec3) {
    img.imageData[(y * img.width + x) * 3    ] = color.X;
    img.imageData[(y * img.width + x) * 3 + 1] = color.Y;
    img.imageData[(y * img.width + x) * 3 + 2] = color.Z;
}

func (img *Image) GetPixel(x,y int) vec3.Vec3 {
    return vec3.Vec3{
        img.imageData[(y * img.width + x) * 3    ],
        img.imageData[(y * img.width + x) * 3 + 1],
        img.imageData[(y * img.width + x) * 3 + 2],
    }
}

func (img *Image) Write(filename string) error {
    writer := imageWriter.New( img.imageData , img.width, img.height)
    return writer.Write(filename);
}

